tests: add functional test for miniscript decaying multisig

mjdietzx opened the pull request:
This is very closely based on [test/functional/wallet_multisig_descriptor_psbt.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_multisig_descriptor_psbt.py) both in code and concept. It should serve as some integration testing for Miniscript descriptors, and also documents a simple multisig that starts as 4-of-4 and decays to 3-of-4, 2-of-4, and finally 1-of-4 at block heights (I think in the real world aligning this to halvenings would be nice).

kevkevinpal commented:
PR description link is routes to 404 can you update to use this https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_multisig_descriptor_psbt.py instead?

kevkevinpal commented:
Concept ACK [2422b90](https://github.com/bitcoin/bitcoin/pull/29156/commits/2422b90978f4ea13ee49954598dc8aef841e36df)

I've done a little code review and so far it is looking good, I think it would be cool if instead of always using the pubkeys in the same order when decaying we could randomly use different signers, I just quickly tested it worked the other way around with this diff (not sure how you feel about adding this randomness)

```
             for m in range(self.M):
-                signers_multisig = participants["multisigs"][m]
+                self.log.info(self.N - 1 - m)
+                signers_multisig = participants["multisigs"][self.N - 1 - m]
                 self._check_psbt(psbt["psbt"], to, value, signers_multisig)
-                signing_wallet = participants["signers"][m]
+                signing_wallet = participants["signers"][self.N - 1 - m]
                 psbt = signing_wallet.walletprocesspsbt(psbt["psbt"])
```

That way we can assert that any of the keys can create the 4/4 3/4 2/4 and 1/4

Also, one thing I am unsure of is if we should include this test in `test/functional/wallet_multisig_descriptor_psbt.py` since they are very similar and maybe we can use common functions but I am not sure what will be the cleanest

mjdietzx commented:
Thank you for the review, and good idea @kevkevinpal!

>  I think it would be cool if instead of always using the pubkeys in the same order when decaying we could randomly use different signers

I added a 2nd commit to this PR with your suggestion. In a similar spirit I improved #29154 (also adding a 2nd commit with your suggestion, and in that case we can also assert that order of xpubs in the multisig descriptors don't matter since it is `sorted`).

> Also, one thing I am unsure of is if we should include this test in test/functional/wallet_multisig_descriptor_psbt.py since they are very similar and maybe we can use common functions but I am not sure what will be the cleanest

I don't think it would be too much trouble to combine them, but I didn't go that route because my opinion is that some code duplication is preferable to trying to generalize/combine these tests. So I went the route of keeping these as two simple independent integration tests. I don't have a strong opinion on this, but usually for test code like this I lean towards approach I took where I duplicate some code to favor simple and easy to explain/delete tests. If more people feel strongly one way or the other I'm happy to change though

kevkevinpal commented:
ACK [7f0be89](https://github.com/bitcoin/bitcoin/pull/29156/commits/7f0be8969e721de69fc352f76db5787836803b76)

michaelfolkson commented:
Concept ACK

A couple of things on first skim read.

1) I think it is probably overkill to create a new functional test file for this. I'm not sure why it can't go in `wallet_miniscript.py`?

2) Miniscript has always been described by its authors/contributors as an extension of descriptors. I personally wouldn't be against adding some Miniscript explanation, examples in `descriptors.md` but I'd want a short explanation of what Miniscript is making it clear that this decaying multisig example is enabled by Miniscript and not just added to the list of descriptor examples. Maybe longer term there should be a `miniscript.md` but for now that seems fine to me.

Eunovo commented:
ACK [7f0be89](https://github.com/bitcoin/bitcoin/pull/29156/commits/7f0be8969e721de69fc352f76db5787836803b76)

Adding an example for this to descriptors.md is really helpful but I think users who are unfamiliar with miniscript and descriptors might benefit from the addition of the fragments, `thresh` and `after` to the `Reference` section. Seeing these fragments used in the examples but not included the reference can introduce confusion.

Eunovo commented:
I tested this locally and I've observed that the test doesn't fail if you modify the external wallet's timelocks, see below:
`external = multisig.getdescriptorinfo(f"wsh(thresh({self.N},pk({f'),s:pk('.join(external_xpubs)}),sln:after(128)))")`
The reason is that the test only sends from the external wallet once using all keys and the transactions that use the `locktimes` send from the internal wallet because that's where the funds are. Hence, changing the timelocks for the external descriptor doesn't break the test.

I'm not sure if we care about this or not but fixing this is pretty simple, we just need to create a transaction that uses 3-of-4 keys and the first timelock and sends from the external wallet. The easiest way to do this is to modify the first transaction in the test. See sample code below:
```
self.generate(self.nodes[0], self.locktimes[0])
self.sync_blocks()
self.log.info("First, make a sending transaction, created using `walletcreatefundedpsbt` (anyone can initiate this)...")
psbt = participants["multisigs"][0].walletcreatefundedpsbt(inputs=[], outputs={to: value}, feeRate=0.00010, locktime=self.locktimes[0])

psbts = []
self.log.info("Now at least M users check the psbt with decodepsbt and (if OK) signs it with walletprocesspsbt...")
for m in random.sample(range(self.M), self.M - 1):
    signers_multisig = participants["multisigs"][m]
    self._check_psbt(psbt["psbt"], to, value, signers_multisig)
    signing_wallet = participants["signers"][m]
    partially_signed_psbt = signing_wallet.walletprocesspsbt(psbt["psbt"])
    psbts.append(partially_signed_psbt["psbt"])

self.log.info("Finally, collect the signed PSBTs with combinepsbt, finalizepsbt, then broadcast the resulting transaction...")
combined = coordinator_wallet.combinepsbt(psbts)
finalized = coordinator_wallet.finalizepsbt(combined)
coordinator_wallet.sendrawtransaction(finalized["hex"])
```

mjdietzx commented:
Thank you @michaelfolkson and @Eunovo for the reviews! I agree with both of your feedback re improving `descriptors.md` docs to accompany this test. I will have a commit addressing your feedback in the coming days.

> I think it is probably overkill to create a new functional test file for this. I'm not sure why it can't go in wallet_miniscript.py

My intention with this test is to provide a "reference example" with the same motivations and flow/architecture as #22067 (and the original issue #21278 it resolved). But taking this one step further with a "decaying" multisig that highlights "Miniscript as an extension of descriptors" and showing the same multisig architecture and signing flows "just work" when we utilize Miniscript.

To your point - the additional raw test coverage this provides is limited, and if anything could be accomplished in `wallet_miniscript.py`. So if a "reference example" that tests end-to-end functionality of a wallet like this at a high-level (and more from a user perspective) isn't valuable then this is overkill and is not needed.

I think once I follow up with docs this intention will be clear. Then we can see if this additional test is useful or overkill.

> modify(ing) the external wallet's timelocks

@Eunovo interesting observation! We could also break up the initial deposit into multiple utxos so that we aren't only spending change after the first transaction. I played around with it locally after thinking through your writeup, but I'm leaning towards not modifying the test to try to assert this failure. Partly because modifying the external descriptor to differ from internal would be a _different wallet_ conceptually than what I have here. Ideally #22838 would be merged and I'd be using those semantics in this test and then it's even more clear that external and internal wallets should only in derivation path

Regardless, it's cool you thought of this. For a similar wallet that uses relative time locks instead of absolute, this would be even weirder. Because the same concept could lead to the change of a spendable utxo being unspendable for a time.

epiccurious commented:
Concept ACK 5023b47e85161a0b35839ce03c7b5d55ff1cd4c1.

achow101 commented:
> This is very closely based on [test/functional/wallet_multisig_descriptor_psbt.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_multisig_descriptor_psbt.py) both in code and concept.

Can you articulate what meaningful difference there is?

mjdietzx commented:
> Can you articulate what meaningful difference there is?

Tests a Miniscript descriptor and asserts that it behaves as expected at different block heights as the multisig's thresh of required signers decreases.

I can't judge how meaningful that is or if it deserves a standalone test. As someone who checks in occasionally I needed to convince myself I can really throw Miniscript into a descriptor and the wallet will work/behave like I'm used to

mjdietzx commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,177 @@
+            self.log.info(f"At block height >= {locktime} this multisig is {self.M}-of-{self.N}")
+            current_height = self.nodes[0].getblock(self.nodes[0].getbestblockhash())['height']
+
+            psbt = participants["multisigs"][0].walletcreatefundedpsbt(inputs=[], outputs={to: value}, feeRate=0.00010, locktime=locktime)
```
An interesting thing I learned with this test was psbt behavior related to `locktime`. If I created a psbt at a blockheight of 800,000 when multisig was 4-of-4 (maybe I want to pre-sign it with a cold wallet and save it for later), and then I wanted to spend it when the multisig decayed to 3-of-4 at blockheight 840,000, then wallet would not process that psbt as `complete` with 3 valid signatures like I expected (bc I guess it's using locktime of when psbt was created to know how to interpret the Miniscript and not current blockheight)

I didn't investigate further or learn why that is.. or if there is a fundamental reason for this behavior or if `walletprocesspsbt` can/should be improved

rkrux replied:
> then wallet would not process that psbt as complete with 3 valid signatures like I expected

Do you mean the 4-signature transaction was rejected?

mjdietzx replied:
I just got confused here. CLTV opcode checks against spending transaction `nLockTime`. So behavior is exactly as expected

> Do you mean the 4-signature transaction was rejected?

No I didn't mean this. 4-signature transaction would never be rejected. But I explained this confusingly so you can ignore my comment above

rkrux commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,176 @@
+                signing_wallet = participants["signers"][m]
+                psbt = signing_wallet.walletprocesspsbt(psbt["psbt"])
+                assert_equal(psbt["complete"], m == self.M - 1)
+            self.log.info("Check that the time-locked transaction is too immature to spend...")
+            assert_equal(current_height >= locktime, False)
+            assert_raises_rpc_error(-26, "non-final", coordinator_wallet.sendrawtransaction, psbt["hex"])
+
+            self.generate(self.nodes[0], locktime)
+            self.sync_blocks()
+            coordinator_wallet.sendrawtransaction(psbt["hex"])
+
+            self.log.info("Check that balances are correct after the transaction has been included in a block.")
```
```
2024-04-13T08:31:05.463000Z TestFramework (INFO): At block height >= 128 this multisig is 3-of-4
2024-04-13T08:31:05.473000Z TestFramework (INFO): Check that the time-locked transaction is too immature to spend...
2024-04-13T08:31:05.614000Z TestFramework (INFO): Check that balances are correct after the transaction has been included in a block.
```

The first time I ran this test, I felt there should be another log message in between that displays something like "Generating ABC blocks to cover for the timelock". Reading the 3rd message immediately after the 2nd one made me think this new message might make what the test is doing more explicit.

mjdietzx replied:
Good point! I added the message like you suggest. Logs now read:
```
2024-04-13T23:05:41.654000Z TestFramework (INFO): At block height >= 128 this multisig is 3-of-4
2024-04-13T23:05:41.674000Z TestFramework (INFO): Check that the time-locked transaction is too immature to spend with 3-of-4 at block height 103...
2024-04-13T23:05:41.675000Z TestFramework (INFO): Generate blocks to reach the time-lock block height 128 and broadcast the transaction...
2024-04-13T23:05:42.888000Z TestFramework (INFO): Check that balances are correct after the transaction has been included in a block.
```

rkrux commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,176 @@
+        assert_approx(participants["multisigs"][0].getbalance(), deposit_amount - value, vspan=0.001)
+        assert_equal(participants["signers"][self.N - 1].getbalance(), value)
+
+        self.log.info("Send more transactions from the multisig as required signers decay, this time with a daisy chained signing flow (one after another in series)!")
```
Most of the code before this one is very similar to `wallet_multisig_descriptor_psbt.py` test - `participants_create_multisigs` is different though.
As pointed out in other reviews as well, I feel this test can go in the same file and some commonalities can be extracted, but I don't think this is blocking though.

mjdietzx replied:
Yeah there is def some duplicated code between the two tests. But I want to focus on making this simple and easy to understand. It's trying to demonstrate/document an example wallet and could be used as a reference. Kind of like a quick start / tutorial

If I were to combine these tests it would become harder for someone to quickly and easily follow, and would lose the purpose of being a quick example. I think it's quite common to have code duplication in these types of tests/examples for that reason and hope this is fine

rkrux reviewed (approved):
tACK [5023b47](https://github.com/bitcoin/bitcoin/pull/29156/commits/5023b47e85161a0b35839ce03c7b5d55ff1cd4c1)

Build and all functional tests successful. I like the clarity and the expressiveness with which this test has been written, and believe this is a good addition. Thanks for coming up with this @mjdietzx!

Suggested adding a log message in the last portion of the test.

mjdietzx commented:
Thanks for the review @rkrux ! 

> Suggested adding a log message in the last portion of the test.

Nice suggestion, I did this just now when I rebased on latest `master`

achow101 commented on doc/descriptors.md:
```diff
@@ -63,6 +63,7 @@ Output descriptors currently support:
 - `wsh(sortedmulti(1,xpub661MyMwAqRbcFW31YEwpkMuc5THy2PSt5bDMsktWQcFF8syAmRUapSCGu8ED9W6oDMSgv6Zz8idoc4a6mr8BDzTJY47LJhkJ8UB7WEGuduB/1/0/*,xpub69H7F5d8KSRgmmdJg2KhpAK8SR3DjMwAdkxj3ZuxV27CprR9LgpeyGmXUbC6wb7ERfvrnKZjXoUmmDznezpbZb7ap6r1D3tgFxHmwMkQTPH/0/0/*))` describes a set of *1-of-2* P2WSH multisig outputs where one multisig key is the *1/0/`i`* child of the first specified xpub and the other  ...
 - `tr(c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5,{pk(fff97bd5755eeea420453a14355235d382f6472f8568a18b2f057a1460297556),pk(e493dbf1c10d80f3581e4904930b1404cc6c13900ee0758474fa94abe8c4cd13)})` describes a P2TR output with the `c6...` x-only pubkey as internal key, and two script paths.
 - `tr(c6047f9441ed7d6d3045406e95c07cd85c778e4b8cef3ca7abac09b95c709ee5,sortedmulti_a(2,2f8bde4d1a07209355b4a7250a5c5128e88b84bddc619ab7cba8d569b240efe4,5cbdf0646e5db4eaa398f365f2ea7a0e3d419b7e0330e39ce92bddedcac4f9bc))` describes a P2TR output with the `c6...` x-only pubkey as internal key, and a single `multi_a` script that needs 2 signatures with 2 specified x-only keys, which will be sorted le ...
+- `wsh(thresh(4,pk([7258e4f9/44h/1h/0h]tpubDCZrkQoEU3845aFKUu9VQBYWZtrTwxMzcxnBwKFCYXHD6gEXvtFcxddCCLFsEwmxQaG15izcHxj48SXg1QS5FQGMBx5Ak6deXKPAL7wauBU/0/*),s:pk([c80b1469/44h/1h/0h]tpubDD3UwwHoNUF4F3Vi5PiUVTc3ji1uThuRfFyBexTSHoAcHuWW2z8qEE2YujegcLtgthr3wMp3ZauvNG9eT9xfJyxXCfNty8h6rDBYU8UU1qq/0/*),s:pk([4e5024fe/44h/1h/0h]tpubDDLrpPymPLSCJyCMLQdmcWxrAWwsqqssm5NdxT2WSdEBPSXNXxwbeKtsHAyXPpLkhUyKovtZ ...
```
In de93b863583921929bc16feb7a0bf5decfbc7ac5 "tests: add functional test for miniscript decaying multisig"

It's not necessary to state where the descriptor comes from.

achow101 commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,176 @@
+        self.add_wallet_options(parser, legacy=False)
+
+    def set_test_params(self):
+        self.num_nodes = 4
```
In de93b863583921929bc16feb7a0bf5decfbc7ac5 "tests: add functional test for miniscript decaying multisig"

Using multiple wallets is preferred over using multiple nodes.

achow101 commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,176 @@
+        # Keep all key origin information (master key fingerprint and all derivation steps) for proper support of hardware devices
+        # See section 'Key origin identification' in 'doc/descriptors.md' for more details...
+        return pkh_descriptor["desc"].split("pkh(")[1].split(")")[0]
+
+    @staticmethod
+    def _check_psbt(psbt, to, value, multisig):
+        """Helper function for any of the N participants to check the psbt with decodepsbt and verify it is OK before signing."""
+        tx = multisig.decodepsbt(psbt)["tx"]
+        amount = 0
+        for vout in tx["vout"]:
+            address = vout["scriptPubKey"]["address"]
+            assert_equal(multisig.getaddressinfo(address)["ischange"], address != to)
+            if address == to:
+                amount += vout["value"]
+        assert_approx(amount, float(value), vspan=0.001)
+
+    def participants_create_multisigs(self, external_xpubs, internal_xpubs):
+        """The multisig is created by importing the following descriptors. The resulting wallet is watch-only and every participant can do this."""
+        for i, node in enumerate(self.nodes):
+            node.createwallet(wallet_name=f"{self.name}_{i}", blank=True, descriptors=True, disable_private_keys=True)
+            multisig = node.get_wallet_rpc(f"{self.name}_{i}")
+            # Spending policy: `thresh(4,pk(key_1),pk(key_2),pk(key_3),pk(key_4),after(t1),after(t2),after(t3))`
+            external = multisig.getdescriptorinfo(f"wsh(thresh({self.N},pk({f'),s:pk('.join(external_xpubs)}),sln:after({f'),sln:after('.join(map(str, self.locktimes))})))")
+            internal = multisig.getdescriptorinfo(f"wsh(thresh({self.N},pk({f'),s:pk('.join(internal_xpubs)}),sln:after({f'),sln:after('.join(map(str, self.locktimes))})))")
+            result = multisig.importdescriptors([
+                {  # receiving addresses (internal: False)
+                    "desc": external["descriptor"],
+                    "active": True,
+                    "internal": False,
+                    "timestamp": "now",
+                },
+                {  # change addresses (internal: True)
+                    "desc": internal["descriptor"],
+                    "active": True,
+                    "internal": True,
+                    "timestamp": "now",
+                },
+            ])
+            assert all(r["success"] for r in result)
+            yield multisig
```
In de93b863583921929bc16feb7a0bf5decfbc7ac5 "tests: add functional test for miniscript decaying multisig"

A lot of this is very similar to/the same as `wallet_multisig_descriptor_psbt.py`. I would prefer if they were combined into one test file. Note that you can (and should) make separate functions for the different test cases so that there is a distinct separation between them, while still using the same utility functions.

mjdietzx replied:
After addressing comments from your review, I made some simplifications and the two tests are now different enough I think this is no longer the case. The tests are still similar in their high-level structure, but otherwise not much code is duplicated (and definitely no complex/deep logic)

achow101 commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,176 @@
+        self.log.info("Check that balances are correct after the transaction has been included in a block.")
+        self.generate(self.nodes[0], 1)
+        assert_approx(participants["multisigs"][0].getbalance(), deposit_amount - value, vspan=0.001)
+        assert_equal(participants["signers"][self.N - 1].getbalance(), value)
```
In de93b863583921929bc16feb7a0bf5decfbc7ac5 "tests: add functional test for miniscript decaying multisig"

This workflow is very similar to what will happen in the loop below. Can they be consolidated?

mjdietzx replied:
I don't think it was ever necessary for me to demonstrate this alternative signing flow as part of this test. So I simplified and removed this code. Now all sends are tested in the single loop

achow101 commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,176 @@
+    def run_test(self):
+        self.M = self.num_nodes
+        self.N = self.num_nodes
+        self.locktimes = [128, 256, 512]  # in real-world align with future halvenings: [840000, 1050000, 1260000]
```
In de93b863583921929bc16feb7a0bf5decfbc7ac5 "tests: add functional test for miniscript decaying multisig"

Mining hundreds of blocks can take a while and is unnecessary for this test. This could run quite a bit faster with locktimes that are closer to each other.

achow101 commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,176 @@
+            assert_equal(current_height >= locktime, False)
+            assert_raises_rpc_error(-26, "non-final", coordinator_wallet.sendrawtransaction, psbt["hex"])
+
+            self.generate(self.nodes[0], locktime)
```
In de93b863583921929bc16feb7a0bf5decfbc7ac5 "tests: add functional test for miniscript decaying multisig"

This massively overshoots the locktime block each time. I think it would be better to mine just enough blocks to reach the locktime height.

achow101 reviewed (commented):
e0b0a385350d8f7a9d00542cc3d14bb47224d889 "tests: miniscript decaying multisig signing order does not matter" could be squashed into de93b863583921929bc16feb7a0bf5decfbc7ac5 "tests: add functional test for miniscript decaying multisig".

mjdietzx commented:
Thank you for the thorough review @achow101! I'll update this PR addressing all of your comments and ping you when I've pushed (I'll need to find a chunk of time to do this so it may be a week or two)

mjdietzx commented:
I've addressed all your comments and rebased/squashed @achow101

> Using multiple wallets is preferred over using multiple nodes.

This point led me to simplify and remove some unnecessary code/checks. When I had first implemented [test/functional/wallet_multisig_descriptor_psbt.py](https://github.com/bitcoin/bitcoin/blob/master/test/functional/wallet_multisig_descriptor_psbt.py) there was feedback/interest in different nodes/participants coordinating together. That led to some additional checks, code, and showing two different signing flows.

In this test I don't think coordination across nodes is necessary or in the scope of what it's trying to show (focused on the number of required signers decaying at the defined block heights). So I made this improvement as you suggested which led to simplifications and removing chunks of now unnecessary code.

RandyMcMillan commented:
Concept ACK 7ddfc6d

rkrux reviewed (commented):
@mjdietzx I will review this PR again soon as I want to see a decaying multi-sig using miniscript example checked in. Would you be willing to rebase this over master to get the CMake build changes? The other option is for me (and other reviewers) to use the older build commands to test this PR locally, which is fine as well but at the cost of a much longer build time and having two building tools in my local. 
Also, it would be nice to ensure this works with the latest build system though there are only python changes.

mjdietzx commented:
@rkrux thanks for the heads up. I didn't expect changes to build system would affect this. I'll rebase asap, also want to bring in achow's [#22838](https://github.com/bitcoin/bitcoin/pull/22838)

mjdietzx commented:
@rkrux I updated to master and rebased. There were no conflicts and this test continues to pass without needing any modifications. This should be ready to merge

Note: I will update to multi-path descriptors now that it's merged in a tiny followup PR. but I don't want to mess with all the prior reviews on this PR so not introducing any new changes

rkrux reviewed (commented):
tACK aae032015367be812154a0e87955e6fd5ea5c192

My comments are mostly nits and few questions.

mjdietzx commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+            # the random sample asserts that any of the signing keys can sign for the 3-of-4,
+            # 2-of-4, and 1-of-4. While this is basic behavior of the miniscript thresh primitive,
+            # it is a critical property of this wallet.
+            for i, m in enumerate(random.sample(range(self.M), self.M)):
+                psbt = signers[m].walletprocesspsbt(psbt["psbt"])
+                assert_equal(psbt["complete"], i == self.M - 1)
```
I don't think it'd hurt to add, but also not necessary because we will be broadcasting this to mempool next and that would revert. if it is simple to add testmempoolaccept though I'll add this as an improvement when I rebase

mjdietzx replied:
I looked into adding this, but `psbt` object does not include the raw tx that we can broadcast (`psbt['hex'`) until the psbt is complete. so it would have been weird/difficult to add this additional check because the interface already prevents it. so i won't add this

mjdietzx commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+
+                self.log.info(f"Generate blocks to reach the time-lock block height {locktime} and broadcast the transaction...")
+                self.generate(self.node, locktime - current_height)
+            else:
```
It's just ensuring that no matter what 4 of 4 signers can always spend. So this is independent of block height. So this is asserted by checking that 4 of 4 can spend before the first lock time. Maybe it's confusing bc it's obvious but I thought it was important to check

rkrux replied:
I do believe it's important to check and if being done at any height less than the first locktime, then indeed all the signers are needed. I got a little thrown off by the word "always" in the middle because I expected a phrase like "All the signers are required to spend before the first locktime".

mjdietzx replied:
I updated the log to what you expected because i think it makes more sense. Changed from "We can always spend with all signers!" to "We can always spend with all signers"

rkrux replied:
I believe you updated it to "All the signers are required to spend before the first locktime" that is easier for me to understand in the given context, thank you.

mjdietzx commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+            self.log.info("Check that balances are correct after the transaction has been included in a block...")
+            self.generate(self.node, 1)
+            assert_approx(multisig.getbalance(), deposit_amount - sent, vspan=0.001)
+            assert_equal(receiver.getbalance(), sent)
```
because the receiver gets the full amount sent, while the sender sent the full amount plus a small txn fee (which is why sender needs to use assert approx to keep it simple)

mjdietzx commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+        assert_equal(len(self.locktimes), self.N - 1)
+
+        self.name = f"{self.M}_of_{self.N}_decaying_multisig"
+        self.log.info(f"Testing a miniscript multisig which starts as 4-of-4 and 'decays' to 3-of-4 at block height {self.locktimes[0]}, 2-of-4 at {self.locktimes[1]}, and finally 1-of-4 at {self.locktimes[2]}...")
```
seems more complicated to me tbh

rkrux replied:
Wanted to avoid hardcoding, feel free to ignore.

mjdietzx commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+        self.M = 4  # starts as 4-of-4
+        self.N = 4
+
+        self.locktimes = [104, 106, 108]  # in real-world align with future halvenings, eg: [840000, 1050000, 1260000]
```
yeah maybe it's over specific. I'll either remove or improve it

mjdietzx replied:
Removed this comment

mjdietzx reviewed (commented):
Thank you for the thorough review @rkrux! I've responded to all your comments. I will @ you when I've rebased and addressed the ones that need it

rkrux commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+        multisig = self.create_multisig(external_xpubs, internal_xpubs)
+
+        self.log.info("Get a mature utxo to send to the multisig...")
+        coordinator_wallet = self.node.get_wallet_rpc(self.node.createwallet(wallet_name=f"coordinator", descriptors=True)["name"])
+        self.generatetoaddress(self.node, 101, coordinator_wallet.getnewaddress())
+
+        self.log.info("Send funds to the multisig's receiving address...")
+        deposit_amount = 6.15
+        coordinator_wallet.sendtoaddress(multisig.getnewaddress(), deposit_amount)
```
I don't think there would be self sends from wallets point of view because in the loop the funds are transferred from the `multisig` wallet to the `signer` wallet, both of which are two different wallets with different balances as I checked.

rkrux commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+
+        self.log.info("Send transactions from the multisig as required signers decay...")
+        sent = 0
+        for locktime in [0] + self.locktimes:
```
I get it now, 0 acts a "default/initial" locktime.

mjdietzx committed:
tests: add functional test for miniscript decaying multisig

This is similar in structure to test/functional/wallet_multisig_descriptor_psbt.py
both in code and concept. It should serve as some integration testing for
Miniscript descriptors, and also documents a simple multisig that starts as 4-of-4
and decays to 3-of-4, 2-of-4, and finally 1-of-4 at block heights (I think in the
real world aligning this to halvenings would be nice).

mjdietzx commented on test/functional/wallet_miniscript_decaying_multisig_descriptor_psbt.py:
```diff
@@ -0,0 +1,137 @@
+            current_height = self.node.getblock(self.node.getbestblockhash())['height']
+
+            amount = 1.5
+            receiver = signers[0]
```
Done

mjdietzx commented:
@maflcko I fixed the lint errors, but CI failed for an unrelated technical reason. Are you able to trigger CI to re-run lint job? @rkrux I've rebased with the improvements based on your review and resolved/responded to your comments in PR

rkrux reviewed (approved):
reACK bb633c9407c46eeadf6fe85e859cea1fed24473f

@mjdietzx Thanks for addressing all the comments!

hodlinator reviewed (approved):
ACK bb633c9407c46eeadf6fe85e859cea1fed24473f

Pretty neat to see thresholds being used in that way - as block heights are reached, each fulfilled locktime contributes to the count for the `thresh`. Was imagining some gigantic repeated `sortedmulti` + `or` + `after` combo (which would not have the key order problem inside of each `sortedmulti`, but would still require correct order of everything else and be prohibitively big).

PR helped me realize how different [Miniscript](https://github.com/bitcoin/bips/blob/master/bip-0379.md) syntax is from [Output Descriptors](https://github.com/bitcoin/bips/blob/master/bip-0380.mediawiki). Miniscript is more of an intermediate representation than I remembered.

Good to see the test attempts premature signing of the PSBTs and verifies failure. It means the less readable "identity" specifiers in the Miniscript are very likely correct. Added the suggested log statement to get the miniscript for the receive addresses, replaced keys+paths with letters to get:
```
thresh(4,pk(A),s:pk(B),s:pk(C),s:pk(D),sln:after(104),sln:after(106),sln:after(108))
```
Put that into the analyzer at https://bitcoin.sipa.be/miniscript/ and got a Bitcoin Script that seems to do the right thing; adding together CHECKSIG results with optional CLTV checks to see if we get 4 in total. Quite magical how the CLTV checks are opted into automatically without having to specify more than the `locktime`-parameter to the `walletcreatefundedpsbt`-RPC.

Was only able to find some **nits** at this point, but PR has gone through a fair amount of changes already (switching to 1 node, only doing PSBTs-signing in 1 way instead of 2).

achow101 commented:
ACK bb633c9407c46eeadf6fe85e859cea1fed24473f

rkrux commented:
Glad to see this PR merged!
